src/core/thread/link_quality.cpp

```diff
@@ -0,0 +1,274 @@
+
+    if (mCount != 0)
+    {
+        average = -(static_cast<int8_t>(mRssAverage >> 3));
+    }
+
+    return average;
+}
+
+uint16_t LinkQualityInfo::GetAverageRssAsEncodedWord(void) const
+{
+    return mRssAverage;
+}
+
+ThreadError LinkQualityInfo::GetAverageRssAsString(char *aCharBuffer, size_t aBufferLen) const
+{
+    ThreadError error = kThreadError_None;
+    int charsWritten = 0;
+
+    if (mCount == 0)
+    {
+        VerifyOrExit(aBufferLen >= sizeof(kUnknownRssString), error = kThreadError_NoBufs);
+
+        strcpy(aCharBuffer, kUnknownRssString);
+    }
+    else
+    {
+        charsWritten = snprintf(aCharBuffer, aBufferLen, "%d.%s dBm", GetAverageRss(),
+                                kLinkQualityDecimalDigitsString[mRssAverage & 0x07]);
+
+        VerifyOrExit(charsWritten >= 0, error = kThreadError_NoBufs);
+
+        VerifyOrExit(static_cast<size_t>(charsWritten) < aBufferLen, error = kThreadError_NoBufs);
+    }
+
+exit:
+    return error;
+}
+
+uint8_t LinkQualityInfo::GetLinkMargin(void) const
+{
+    return ConvertRssToLinkMargin(GetAverageRss());
+}
+
+uint8_t LinkQualityInfo::GetLinkQuality(void)
+{
+    UpdateLinkQuality();
+
+    return mLinkQuality;
+}
+
+void LinkQualityInfo::UpdateLinkQuality(void)
+{
+    if (mCount > 1)
+    {
+        mLinkQuality = CalculateLinkQuality(GetLinkMargin(), mLinkQuality);
+    }
+    else
+    {
+        mLinkQuality = CalculateLinkQuality(GetLinkMargin(), kNoLastLinkQualityValue);
+    }
+}
+
+int8_t LinkQualityInfo::GetAverageNoiseFloor(void)
+{
+    int8_t averageNoiseFloor = sNoiseFloorAverage.GetAverageRss();
+
+    if (averageNoiseFloor == kUnknownRss)
+    {
+        averageNoiseFloor = kDefaultNoiseFloor;
+    }
+
+    return averageNoiseFloor;
+}
+
+ThreadError LinkQualityInfo::AddNoiseFloor(int8_t aNoiseFloor)
+{
+    return sNoiseFloorAverage.AddRss(aNoiseFloor);
+}
+
+void LinkQualityInfo::ClearNoiseFloorAverage(void)
+{
+    sNoiseFloorAverage.Clear();
+}
+
+uint8_t LinkQualityInfo::ConvertRssToLinkMargin(int8_t anRss)
+{
+    int8_t linkMargin = anRss - GetAverageNoiseFloor();
+
+    if (linkMargin < 0)
+    {
+        linkMargin = 0;
+    }
+
+    return static_cast<uint8_t>(linkMargin);
+}
+
+uint8_t LinkQualityInfo::ConvertLinkMarginToLinkQuality(uint8_t aLinkMargin)
+{
+    return CalculateLinkQuality(aLinkMargin, kNoLastLinkQualityValue);
+}
+
+uint8_t LinkQualityInfo::ConvertRssToLinkQuality(int8_t anRss)
+{
+    return ConvertLinkMarginToLinkQuality(ConvertRssToLinkMargin(anRss));
+}
+
+uint8_t LinkQualityInfo::CalculateLinkQuality(uint8_t aLinkMargin, uint8_t aLastLinkQuality)
+{
+    uint8_t threshold1, threshold2, threshold3;
+    uint8_t linkQuality = 0;
+
+    threshold1 = kLinkMarginThresholdForLinkQuality1;
+    threshold2 = kLinkMarginThresholdForLinkQuality2;
+    threshold3 = kLinkMarginThresholdForLinkQuality3;
+
+    // Apply the hysteresis threshold based on the last link quality value.
+
+    switch (aLastLinkQuality)
```

Comment:
Nice! Makes more sense to apply the hysteresis on both sides (up and down). When I read the spec, this was what I initially expected +1/-1 dB around the threshold, but the example used +12 dB for move to 2, so I just applied the hysteresis on the upside. I guess we still need some memory (last value) or  to determine if we are moving up or down, right? 

I think I keep the current implementation for now... Should be easy to change it when the spec is updated. Makes sense?